File: src/interactions/fill.js
import { TIMING } from '../core/constants.js';
import { isNavigationError } from '../core/navigation-safety.js';
import { waitForLocatorOrElement } from '../elements/locators.js';
import { scrollIntoViewIfNeeded } from './scroll.js';
import { clickElement } from './click.js';
import { getInputValue } from '../elements/content.js';
import { createEngineAdapter } from '../core/engine-adapter.js';

/**
 * Default verification function for fill operations.
 * Verifies that the filled text matches expected text.
 * @param {Object} options - Verification options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Object} options.locatorOrElement - Element that was filled
 * @param {string} options.expectedText - Text that should be in the element
 * @returns {Promise<{verified: boolean, actualValue: string}>}
 */
export async function defaultFillVerification(options = {}) {
  const { page, engine, locatorOrElement, expectedText } = options;

  try {
    const actualValue = await getInputValue({ page, engine, locatorOrElement });
    // Verify that the value contains the expected text (handles cases where value may have formatting)
    const verified = actualValue === expectedText || actualValue.includes(expectedText);
    return { verified, actualValue };
  } catch (error) {
    if (isNavigationError(error)) {
      return { verified: false, actualValue: '', navigationError: true };
    }
    throw error;
  }
}

/**
 * Verify fill operation with retry logic
 * @param {Object} options - Verification options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type
 * @param {Object} options.locatorOrElement - Element to verify
 * @param {string} options.expectedText - Expected text value
 * @param {Function} options.verifyFn - Custom verification function (optional, defaults to defaultFillVerification)
 * @param {number} options.timeout - Verification timeout in ms (default: TIMING.VERIFICATION_TIMEOUT)
 * @param {number} options.retryInterval - Interval between retries (default: TIMING.VERIFICATION_RETRY_INTERVAL)
 * @param {Function} options.log - Logger instance
 * @returns {Promise<{verified: boolean, actualValue: string, attempts: number}>}
 */
export async function verifyFill(options = {}) {
  const {
    page,
    engine,
    locatorOrElement,
    expectedText,
    verifyFn = defaultFillVerification,
    timeout = TIMING.VERIFICATION_TIMEOUT,
    retryInterval = TIMING.VERIFICATION_RETRY_INTERVAL,
    log = { debug: () => {} },
  } = options;

  const startTime = Date.now();
  let attempts = 0;
  let lastResult = { verified: false, actualValue: '' };

  while (Date.now() - startTime < timeout) {
    attempts++;
    lastResult = await verifyFn({
      page,
      engine,
      locatorOrElement,
      expectedText,
    });

    if (lastResult.verified) {
      log.debug(() => `✅ Fill verification succeeded after ${attempts} attempt(s)`);
      return { ...lastResult, attempts };
    }

    if (lastResult.navigationError) {
      log.debug(() => '⚠️  Navigation detected during fill verification');
      return { ...lastResult, attempts };
    }

    // Wait before next retry
    await new Promise(resolve => setTimeout(resolve, retryInterval));
  }

  log.debug(() => `❌ Fill verification failed after ${attempts} attempts. Expected: "${expectedText}", Got: "${lastResult.actualValue}"`);
  return { ...lastResult, attempts };
}

/**
 * Check if an input element is empty
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Object} options.locatorOrElement - Element or locator to check
 * @param {Object} options.adapter - Engine adapter (optional, will be created if not provided)
 * @returns {Promise<boolean>} - True if empty, false if has content (returns true on navigation)
 */
export async function checkIfElementEmpty(options = {}) {
  const { page, engine, locatorOrElement, adapter: providedAdapter } = options;

  if (!locatorOrElement) {
    throw new Error('locatorOrElement is required in options');
  }

  // Add defensive check for page parameter
  if (!page && !providedAdapter) {
    const availableKeys = Object.keys(options).join(', ');
    throw new Error(`checkIfElementEmpty: page is required in options when adapter is not provided. Available option keys: [${availableKeys}]. This indicates the 'page' parameter was not passed correctly from the calling function.`);
  }

  try {
    const adapter = providedAdapter || createEngineAdapter(page, engine);
    const currentValue = await adapter.getInputValue(locatorOrElement);
    return !currentValue || currentValue.trim() === '';
  } catch (error) {
    if (isNavigationError(error)) {
      console.log('⚠️  Navigation detected during checkIfElementEmpty, returning true');
      return true;
    }
    throw error;
  }
}

/**
 * Perform fill/type operation on an element (low-level)
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Object} options.locatorOrElement - Element or locator to fill
 * @param {string} options.text - Text to fill
 * @param {boolean} options.simulateTyping - Whether to simulate typing (default: true)
 * @param {boolean} options.verify - Whether to verify the fill operation (default: true)
 * @param {Function} options.verifyFn - Custom verification function (optional)
 * @param {number} options.verificationTimeout - Verification timeout in ms (default: TIMING.VERIFICATION_TIMEOUT)
 * @param {Function} options.log - Logger instance (optional)
 * @param {Object} options.adapter - Engine adapter (optional, will be created if not provided)
 * @returns {Promise<{filled: boolean, verified: boolean, actualValue?: string}>}
 */
export async function performFill(options = {}) {
  const {
    page,
    engine,
    locatorOrElement,
    text,
    simulateTyping = true,
    verify = true,
    verifyFn,
    verificationTimeout = TIMING.VERIFICATION_TIMEOUT,
    log = { debug: () => {} },
    adapter: providedAdapter,
  } = options;

  if (!text) {
    throw new Error('text is required in options');
  }

  if (!locatorOrElement) {
    throw new Error('locatorOrElement is required in options');
  }

  // Add defensive check for page parameter
  if (!page && !providedAdapter) {
    const availableKeys = Object.keys(options).join(', ');
    throw new Error(`performFill: page is required in options when adapter is not provided. Available option keys: [${availableKeys}]. This indicates the 'page' parameter was not passed correctly from the calling function.`);
  }

  try {
    const adapter = providedAdapter || createEngineAdapter(page, engine);
    if (simulateTyping) {
      await adapter.type(locatorOrElement, text);
    } else {
      await adapter.fill(locatorOrElement, text);
    }

    // Verify fill if requested
    if (verify) {
      const verificationResult = await verifyFill({
        page,
        engine,
        locatorOrElement,
        expectedText: text,
        verifyFn,
        timeout: verificationTimeout,
        log,
      });

      if (!verificationResult.verified) {
        log.debug(() => `⚠️  Fill verification failed: expected "${text}", got "${verificationResult.actualValue}"`);
      }

      return {
        filled: true,
        verified: verificationResult.verified,
        actualValue: verificationResult.actualValue,
      };
    }

    return { filled: true, verified: true };
  } catch (error) {
    if (isNavigationError(error)) {
      console.log('⚠️  Navigation detected during performFill, recovering gracefully');
      return { filled: false, verified: false };
    }
    throw error;
  }
}

/**
 * Fill a textarea with text (high-level with checks and scrolling)
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {string} options.engine - Engine type ('playwright' or 'puppeteer')
 * @param {Function} options.wait - Wait function
 * @param {Function} options.log - Logger instance
 * @param {string|Object} options.selector - CSS selector or Playwright Locator
 * @param {string} options.text - Text to fill
 * @param {boolean} options.checkEmpty - Only fill if empty (default: true)
 * @param {boolean} options.scrollIntoView - Scroll into view (default: true)
 * @param {boolean} options.simulateTyping - Simulate typing vs direct fill (default: true)
 * @param {number} options.timeout - Timeout in ms (default: TIMING.DEFAULT_TIMEOUT)
 * @param {boolean} options.verify - Whether to verify the fill operation (default: true)
 * @param {Function} options.verifyFn - Custom verification function (optional, uses defaultFillVerification if not provided)
 * @param {number} options.verificationTimeout - Verification timeout in ms (default: TIMING.VERIFICATION_TIMEOUT)
 * @returns {Promise<{filled: boolean, verified: boolean, skipped: boolean, actualValue?: string}>}
 *   - filled: true if fill operation was attempted
 *   - verified: true if fill was verified successfully (only meaningful if filled is true)
 *   - skipped: true if element already had content and checkEmpty was true
 * @throws {Error} - If selector or text is missing, or if operation fails (except navigation)
 */
export async function fillTextArea(options = {}) {
  const {
    page,
    engine,
    wait,
    log,
    selector,
    text,
    checkEmpty = true,
    scrollIntoView: shouldScroll = true,
    simulateTyping = true,
    timeout = TIMING.DEFAULT_TIMEOUT,
    verify = true,
    verifyFn,
    verificationTimeout = TIMING.VERIFICATION_TIMEOUT,
  } = options;

  // Defensive check: Validate that page parameter is present
  if (!page) {
    const availableKeys = Object.keys(options).join(', ');
    throw new Error(`fillTextArea: page is required in options. Available option keys: [${availableKeys}]. This indicates the 'page' parameter was not passed correctly from the calling function (bindings layer).`);
  }

  if (!selector || !text) {
    throw new Error('fillTextArea: selector and text are required in options');
  }

  try {
    // Get locator/element and wait for it to be visible (unified for both engines)
    const locatorOrElement = await waitForLocatorOrElement({ page, engine, selector, timeout });

    // Check if empty (if requested)
    if (checkEmpty) {
      const isEmpty = await checkIfElementEmpty({ page, engine, locatorOrElement });
      if (!isEmpty) {
        const currentValue = await getInputValue({ page, engine, locatorOrElement });
        log.debug(() => `🔍 [VERBOSE] Textarea already has content, skipping: "${currentValue.substring(0, 30)}..."`);
        return { filled: false, verified: false, skipped: true, actualValue: currentValue };
      }
    }

    // Scroll into view (if requested and needed)
    if (shouldScroll) {
      await scrollIntoViewIfNeeded({ page, engine, wait, log, locatorOrElement, behavior: 'smooth' });
    }

    // Click the element (prevent auto-scroll if scrollIntoView is disabled)
    const clicked = await clickElement({ page, engine, log, locatorOrElement, noAutoScroll: !shouldScroll });
    if (!clicked) {
      return { filled: false, verified: false, skipped: false }; // Navigation occurred
    }

    // Fill the text with verification
    const fillResult = await performFill({
      page,
      engine,
      locatorOrElement,
      text,
      simulateTyping,
      verify,
      verifyFn,
      verificationTimeout,
      log,
    });

    if (!fillResult.filled) {
      return { filled: false, verified: false, skipped: false }; // Navigation occurred
    }

    log.debug(() => `🔍 [VERBOSE] Filled textarea with text: "${text.substring(0, 50)}..."`);

    if (fillResult.verified) {
      log.debug(() => `✅ Fill verification passed`);
    } else {
      log.debug(() => `⚠️  Fill verification failed: expected "${text}", got "${fillResult.actualValue}"`);
    }

    return {
      filled: true,
      verified: fillResult.verified,
      skipped: false,
      actualValue: fillResult.actualValue,
    };
  } catch (error) {
    if (isNavigationError(error)) {
      console.log('⚠️  Navigation detected during fillTextArea, recovering gracefully');
      return { filled: false, verified: false, skipped: false };
    }
    throw error;
  }
}

File: rust/src/interactions/fill.rs
//! Fill operations for browser automation.
//!
//! This module provides utilities for filling form elements
//! with verification support.

use crate::core::constants::TIMING;
use crate::core::engine::{EngineAdapter, EngineError, FillVerificationResult};
use crate::core::navigation::is_navigation_error;
use crate::elements::content::is_element_empty;
use crate::interactions::click::click_element;
use crate::interactions::scroll::{scroll_into_view_if_needed, ScrollOptions};
use std::time::{Duration, Instant};

/// Options for fill operations.
#[derive(Debug, Clone)]
pub struct FillOptions {
    /// Whether to scroll the element into view before filling.
    pub scroll_into_view: bool,
    /// Whether to simulate typing (character by character).
    pub simulate_typing: bool,
    /// Only fill if the element is empty.
    pub check_empty: bool,
    /// Whether to verify the fill operation.
    pub verify: bool,
    /// Timeout for verification.
    pub verification_timeout: Duration,
    /// Interval between verification retries.
    pub verification_retry_interval: Duration,
    /// Timeout for the overall operation.
    pub timeout: Duration,
}

impl Default for FillOptions {
    fn default() -> Self {
        Self {
            scroll_into_view: true,
            simulate_typing: true,
            check_empty: true,
            verify: true,
            verification_timeout: TIMING.verification_timeout,
            verification_retry_interval: TIMING.verification_retry_interval,
            timeout: TIMING.default_timeout,
        }
    }
}

/// Result of a fill operation.
#[derive(Debug, Clone)]
pub struct FillResult {
    /// Whether the fill was performed.
    pub filled: bool,
    /// Whether the fill was verified as successful.
    pub verified: bool,
    /// Whether the fill was skipped (element had content).
    pub skipped: bool,
    /// The actual value after filling.
    pub actual_value: Option<String>,
}

impl FillResult {
    /// Create a successful fill result.
    pub fn success(actual_value: String) -> Self {
        Self {
            filled: true,
            verified: true,
            skipped: false,
            actual_value: Some(actual_value),
        }
    }

    /// Create a result indicating fill was skipped.
    pub fn skipped(actual_value: String) -> Self {
        Self {
            filled: false,
            verified: false,
            skipped: true,
            actual_value: Some(actual_value),
        }
    }

    /// Create a failed fill result.
    pub fn failed() -> Self {
        Self {
            filled: false,
            verified: false,
            skipped: false,
            actual_value: None,
        }
    }
}

/// Verify a fill operation by checking the element's value.
///
/// # Arguments
///
/// * `adapter` - The engine adapter to use
/// * `selector` - The CSS selector for the element
/// * `expected_text` - The expected text value
/// * `options` - Fill options
///
/// # Returns
///
/// The verification result
pub async fn verify_fill(
    adapter: &dyn EngineAdapter,
    selector: &str,
    expected_text: &str,
    options: &FillOptions,
) -> Result<FillVerificationResult, EngineError> {
    let start_time = Instant::now();
    let mut attempts = 0u32;

    while start_time.elapsed() < options.verification_timeout {
        attempts += 1;

        let actual_value = match adapter.input_value(selector).await {
            Ok(Some(value)) => value,
            Ok(None) => String::new(),
            Err(e) if is_navigation_error(&e.to_string()) => {
                return Ok(FillVerificationResult {
                    verified: false,
                    actual_value: String::new(),
                    attempts,
                });
            }
            Err(e) => return Err(e),
        };

        // Verify that the value contains the expected text
        let verified = actual_value == expected_text || actual_value.contains(expected_text);

        if verified {
            return Ok(FillVerificationResult {
                verified: true,
                actual_value,
                attempts,
            });
        }

        tokio::time::sleep(options.verification_retry_interval).await;
    }

    // Final check
    let actual_value = adapter.input_value(selector).await?.unwrap_or_default();

    Ok(FillVerificationResult {
        verified: false,
        actual_value,
        attempts,
    })
}

/// Fill an input element with text (low-level operation).
///
/// # Arguments
///
/// * `adapter` - The engine adapter to use
/// * `selector` - The CSS selector for the element
/// * `text` - The text to fill
/// * `options` - Fill options
///
/// # Returns
///
/// The result of the fill operation
pub async fn perform_fill(
    adapter: &dyn EngineAdapter,
    selector: &str,
    text: &str,
    options: &FillOptions,
) -> Result<FillResult, EngineError> {
    // Perform the fill operation
    if options.simulate_typing {
        match adapter.type_text(selector, text).await {
            Ok(_) => {}
            Err(e) if is_navigation_error(&e.to_string()) => {
                return Ok(FillResult::failed());
            }
            Err(e) => return Err(e),
        }
    } else {
        match adapter.fill(selector, text).await {
            Ok(_) => {}
            Err(e) if is_navigation_error(&e.to_string()) => {
                return Ok(FillResult::failed());
            }
            Err(e) => return Err(e),
        }
    }

    // Verify if requested
    if options.verify {
        let verification = verify_fill(adapter, selector, text, options).await?;
        Ok(FillResult {
            filled: true,
            verified: verification.verified,
            skipped: false,
            actual_value: Some(verification.actual_value),
        })
    } else {
        Ok(FillResult {
            filled: true,
            verified: true,
            skipped: false,
            actual_value: None,
        })
    }
}

/// Fill a text area with text (high-level with checks and scrolling).
///
/// This function handles:
/// - Checking if the element is empty (if requested)
/// - Scrolling the element into view
/// - Clicking to focus
/// - Filling the text
/// - Verifying the fill
///
/// # Arguments
///
/// * `adapter` - The engine adapter to use
/// * `selector` - The CSS selector for the element
/// * `text` - The text to fill
/// * `options` - Fill options
///
/// # Returns
///
/// The result of the fill operation
pub async fn fill_text_area(
    adapter: &dyn EngineAdapter,
    selector: &str,
    text: &str,
    options: &FillOptions,
) -> Result<FillResult, EngineError> {
    // Check if empty (if requested)
    if options.check_empty {
        let is_empty = is_element_empty(adapter, selector).await?;
        if !is_empty {
            let current_value = adapter.input_value(selector).await?.unwrap_or_default();
            return Ok(FillResult::skipped(current_value));
        }
    }

    // Scroll into view if requested
    if options.scroll_into_view {
        let scroll_options = ScrollOptions::default();
        match scroll_into_view_if_needed(adapter, selector, &scroll_options).await {
            Ok(_) => {}
            Err(e) if is_navigation_error(&e.to_string()) => {
                return Ok(FillResult::failed());
            }
            Err(e) => return Err(e),
        }
    }

    // Click to focus
    let click_options = crate::interactions::click::ClickOptions {
        scroll_into_view: false, // Already scrolled
        verify: false,           // Don't need to verify the focus click
        ..Default::default()
    };

    match click_element(adapter, selector, &click_options).await {
        Ok(_) => {}
        Err(e) if is_navigation_error(&e.to_string()) => {
            return Ok(FillResult::failed());
        }
        Err(e) => return Err(e),
    }

    // Perform the fill
    perform_fill(adapter, selector, text, options).await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fill_options_default() {
        let options = FillOptions::default();
        assert!(options.scroll_into_view);
        assert!(options.simulate_typing);
        assert!(options.check_empty);
        assert!(options.verify);
    }

    #[test]
    fn fill_result_success() {
        let result = FillResult::success("filled text".to_string());
        assert!(result.filled);
        assert!(result.verified);
        assert!(!result.skipped);
        assert_eq!(result.actual_value, Some("filled text".to_string()));
    }

    #[test]
    fn fill_result_skipped() {
        let result = FillResult::skipped("existing text".to_string());
        assert!(!result.filled);
        assert!(!result.verified);
        assert!(result.skipped);
        assert_eq!(result.actual_value, Some("existing text".to_string()));
    }

    #[test]
    fn fill_result_failed() {
        let result = FillResult::failed();
        assert!(!result.filled);
        assert!(!result.verified);
        assert!(!result.skipped);
        assert!(result.actual_value.is_none());
    }
}
